File: HackerRank/Dictionaries_and_Hashmaps/Java/twoStrings.java
import java.util.Hashtable;

public class twoStrings {
    public static void main(String[] args){
        /* I attempted this problem in python first
        so my thoughts should be in twoStrings.py
        the code below is setup for the function */

        String s1 = "hello";
        String s2 = "world";

        System.out.println("Do " + s1 + " and " + s2 + " contain any common substrings?");
        System.out.println(twoStringsProblem(s1, s2));
    }

    // Complete the twoStrings function below.
    static String twoStringsProblem(String s1, String s2) {
        /* ntoes: */

        // create a hashtable to hold all the letters that appear in s1
        Hashtable<Character, Boolean> s1Hash = new Hashtable<>();

        for(int i = 0; i < s1.length(); i++){

            if(s1Hash.size() == 26){
                return "YES"; // if every letter in the alphabet is in s1, both strings must have a substring in common
            }

            s1Hash.putIfAbsent(s1.charAt(i), true); // true here just means that the letter is in s1
        }

        // check the letters s2
        for(int i = 0; i < s2.length(); i++){
            if(s1Hash.get(s2.charAt(i)) == true){
                // this part needs some work
                return "YES";
            }
        }

        return "NO"; // after going through the whole string, there weren't any matches
    }
}

File: HackerRank/Interview_Prep/Python/twoStrings.py
''' to sumarise, I want to find out if s1 and s2 share
a common substring. That means as soon as I find 1 (even
if there are multiple) I can return YES '''

import re

# Complete the twoStrings function below.
def twoStrings(s1, s2):
    ''' things to keep in mind...
    strings can be of length 10^5 (100000)
    All I need to do is check that the two strings
    contain one letter that's the same in both
    '''

    #create a dictionary for the first string
    sDict = {}

    # add each letter that occurs in s1 to dictionary
    for char in s1:
        if len(sDict) == 26:
            return 'YES' # I should be able to return 'YES' here since s1 contains every letter of the alphabet at this point
        elif char not in sDict.keys():
            sDict[char] = 1

    # now check if s2 contains the
    for char in s2:
        if char in sDict.keys():
            return 'YES'

    # if we get to this point, s1 and s2 have no substring in common
    return 'NO'

def main():

    s1 = 'abcdefghijklmnopqrstuvw'
    s2 = 'xzya'

    print(twoStrings(s1, s2))

main()